File: twitter-collect/src/main/java/geomesa/example/twitter/collect/Runner.java
/*
 * Copyright 2014 Commonwealth Computer Research, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package geomesa.example.twitter.collect;

import com.beust.jcommander.JCommander;
import com.beust.jcommander.Parameter;
import com.beust.jcommander.ParameterException;
import org.apache.log4j.BasicConfigurator;
import org.apache.log4j.Level;
import org.apache.log4j.Logger;

import java.io.File;

public class Runner {
    private static final Logger log = Logger.getLogger(Runner.class);

    public static void main(String[] args) throws Exception {
        // configure log4j
        BasicConfigurator.configure();
        Logger.getRootLogger().setLevel(Level.INFO);

        // parse args
        final MyArgs clArgs = new MyArgs();
        final JCommander jc = new JCommander(clArgs);
        try {
            jc.parse(args);
        } catch (ParameterException e) {
            log.info("Error parsing arguments: " + e.getMessage());
            jc.usage();
            System.exit(-1);
        }

        final File outputDir = new File(clArgs.outputDir);
        if(outputDir.exists()) {
           if(!outputDir.isDirectory()){
               log.error("Output Directory "+outputDir.getPath() + " is not a directory");
               System.exit(-1);
           }
        }
        else {
            outputDir.mkdirs();
        }

        // collect tweets
        final TwitterStreamCollector collector = new TwitterStreamCollector(
                outputDir,
                clArgs.consumerKey,
                clArgs.consumerSecret,
                clArgs.token,
                clArgs.secret);

        collector.collect2();
    }

    static class MyArgs {
        @Parameter(names = {"--consumerKey", "-ck"}, description = "Twitter consumer key", required = true)
        String consumerKey;

        @Parameter(names = {"--consumerSecret", "-cs"}, description = "Twitter consumer secret", required = true)
        String consumerSecret;

        @Parameter(names = {"--token", "-t"}, description = "Twitter token", required = true)
        String token;

        @Parameter(names = {"--secret", "-s"}, description = "Twitter secret", required = true)
        String secret;

        @Parameter(names = {"--outputDir", "-o"}, description = "Output Directory", required = true)
        String outputDir;
    }
}

File: twitter-spark/src/main/scala/geomesa/example/twitter/spark/Runner.scala
/*
 * Copyright 2014 Commonwealth Computer Research, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package geomesa.example.twitter.spark

import java.text.SimpleDateFormat

import org.apache.hadoop.conf.Configuration
import org.apache.spark.{SparkConf, SparkContext}
import org.geotools.data.{DataStoreFinder, Query}
import org.geotools.factory.CommonFactoryFinder
import org.locationtech.geomesa.accumulo.data.AccumuloDataStore
import org.locationtech.geomesa.compute.spark.GeoMesaSpark

object Runner {
  def main(args: Array[String]) {

    if (args.length != 1) {
      println("1 arg: feature name")
    }
    val feature = args(0)

    import scala.collection.JavaConversions._

    // Get a handle to the data store
    val params = Map(
      "instanceId" -> "myinstance",
      "zookeepers" -> "zoo1,zoo2,zoo3",
      "user"       -> "user",
      "password"   -> "password",
      "tableName"  -> "geomesa_catalog")

    val ds = DataStoreFinder.getDataStore(params).asInstanceOf[AccumuloDataStore]

    // Construct a CQL query to filter by bounding box (set bbox based on data we collected)
    val ff = CommonFactoryFinder.getFilterFactory2
    val f = ff.bbox("geom", -90.32023, 38.72009, -90.23957, 38.77019, "EPSG:4326")
    val q = new Query(feature, f)

    // Configure Spark
    val conf = new Configuration
    val sconf = GeoMesaSpark.init(new SparkConf(true), ds)
    val sc = new SparkContext(sconf)

    // Create an RDD from a query
    val queryRDD = org.locationtech.geomesa.compute.spark.GeoMesaSpark.rdd(conf, sc, params, q, None)

    // Convert RDD[SimpleFeature] to RDD[(String, SimpleFeature)] where the first
    // element of the tuple is the date to the day resolution
    val dayAndFeature = queryRDD.mapPartitions { iter =>
      val df = new SimpleDateFormat("yyyyMMdd")
      val ff = CommonFactoryFinder.getFilterFactory2
      val exp = ff.property("dtg")
      iter.map { f => (df.format(exp.evaluate(f).asInstanceOf[java.util.Date]), 1) }
    }

    // Group the results by day
    val groupedByDay = dayAndFeature.groupBy { case (date, count) => date }

    // Count the number of features in each day
    val countByDay = groupedByDay.map { case (date, iter) => (date, iter.size) }

    // Collect the results and print
    countByDay.collect().foreach(println)
  }
}
